src/wallet/wallet_tx_builder.cpp

```diff
@@ -0,0 +1,751 @@
+                        sendFromAccount,
+                        allowedChangeTypes({ReceiverType::P2PKH}));
+                if (sendTo.has_value()) {
+                    resolvedPayments.AddPayment(
+                            ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                    changeAddr = sendTo.value();
+                }
+            },
+            [&](const CScriptID& scriptId) {
+                auto sendTo = pwalletMain->GenerateChangeAddressForAccount(
+                        sendFromAccount,
+                        allowedChangeTypes({ReceiverType::P2SH}));
+                assert(sendTo.has_value());
+                resolvedPayments.AddPayment(
+                        ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                changeAddr = sendTo.value();
+            },
+            [&](const libzcash::SproutPaymentAddress& addr) {
+                // for Sprout, we return change to the originating address using the tx builder.
+                changeAddr = addr;
+            },
+            [&](const libzcash::SproutViewingKey& vk) {
+                // for Sprout, we return change to the originating address using the tx builder.
+                changeAddr = vk.address();
+            },
+            [&](const libzcash::SaplingPaymentAddress& addr) {
+                // for Sapling, if using a legacy address, return change to the
+                // originating address; otherwise return it to the Sapling internal
+                // address corresponding to the UFVK.
+                std::optional<RecipientAddress> sendTo;
+                if (sendFromAccount == ZCASH_LEGACY_ACCOUNT) {
+                    sendTo = addr;
+                } else {
+                    sendTo = pwalletMain->GenerateChangeAddressForAccount(
+                            sendFromAccount,
+                            allowedChangeTypes({ReceiverType::Sapling}));
+                }
+                assert(sendTo.has_value());
+                resolvedPayments.AddPayment(
+                        ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                changeAddr = sendTo.value();
+            },
+            [&](const libzcash::SaplingExtendedFullViewingKey& fvk) {
+                // for Sapling, if using a legacy address, return change to the
+                // originating address; otherwise return it to the Sapling internal
+                // address corresponding to the UFVK.
+                std::optional<RecipientAddress> sendTo;
+                if (sendFromAccount == ZCASH_LEGACY_ACCOUNT) {
+                    sendTo = fvk.DefaultAddress();
+                } else {
+                    sendTo = pwalletMain->GenerateChangeAddressForAccount(
+                            sendFromAccount,
+                            allowedChangeTypes({ReceiverType::Sapling}));
+                }
```

Comment:
Similar question here: IIUC a Sapling extfvk should always be associated with the legacy account.

Reply:
Similar response 😁

I’m just going to consolidate on that case, then apply the conclusion across everything.